xapian-letor/api/featurelist.cc

```diff
@@ -0,0 +1,144 @@
+    feature.push_back(new TfDoclenFeature());
+    feature.push_back(new IdfFeature());
+    feature.push_back(new CollTfCollLenFeature());
+    feature.push_back(new TfIdfDoclenFeature());
+    feature.push_back(new TfDoclenCollTfCollLenFeature());
+}
+
+FeatureList::FeatureList(const std::vector<Feature*> & f) {
+
+    std::cout << "Initialising CUSTOM list of features." << endl;
+
+    feature = f;
+}
+
+FeatureList::~FeatureList() {
+    for (std::vector<Feature*>::iterator it = feature.begin() ; it != feature.end(); ++it) {
+        delete (*it);
+    }
+    feature.clear();
+}
+
+void
+FeatureList::normalise(std::vector<FeatureVector> & fvec) {
+
+    // find the max value for each feature gpr all the FeatureVectors in the vector.
+    int num_features = fvec[0].get_fcount();
+    double temp = 0.0;
+    double max[num_features];
+
+    for(int i=0; i<num_features; ++i)
+        max[i] = 0.0;
+
+    int num_fv = fvec.size();
+    for(int i=0; i < num_fv; ++i) {
+        for(int j=0; j<num_features; ++j) {
+            double fval = fvec[i].get_fvals()[j];
+            if (max[j] < fval)
+                max[j] = fval;
+        }
+    }
+
+    /* We have the maximum value of each feature overall.
+       Now we need to normalize each feature value of a
+       featureVector by dividing it by the corresponding max of the feature value
+    */
+
+    for(int i=0; i < num_fv; ++i) {
+        for(int j=0; j<num_features; ++j) {
+            temp = fvec[i].get_feature_value(j);
+            temp /= max[j];
+            if (max[j] == 0) // Skip if dividing by zero
+                continue;
+            fvec[i].set_feature_value(j, temp);
+        }
+    }
+}
+
+std::vector<FeatureVector>
+FeatureList::create_feature_vectors(const Xapian::MSet & mset, const Xapian::Query & letor_query,
+                                    const Xapian::Database & letor_db)
+{
+    std::vector<FeatureVector> fvec;
+
+    for (Xapian::MSetIterator i = mset.begin(); i != mset.end(); ++i) {
+
+        Xapian::Document doc = i.get_document();
+
+        std::vector<double> fVals;
+
+        for (std::vector<Feature*>::iterator it = feature.begin() ; it != feature.end(); ++it) {
+
+            (*it)->set_database(letor_db);
+            (*it)->set_query(letor_query);
+            (*it)->set_doc(doc);
+
+            fVals.push_back((*it)->get_fval('t')); // for title
+            fVals.push_back((*it)->get_fval('b')); // for body
+            fVals.push_back((*it)->get_fval('w')); // for whole
+
+        }
+
+        double wt = i.get_weight();
+
+        /// Weight is added as a feature by default.
+        fVals.push_back(wt);
+
+        Xapian::docid did = doc.get_docid();
+
+        Xapian::FeatureVector fv = create_feature_vector(fVals, did);
+
+        fvec.push_back(fv);
+
+    }
+    normalise(fvec);
+    return fvec;
+}
+
+Xapian::FeatureVector
+FeatureList::create_feature_vector(std::vector<double> & fvals, Xapian::docid & did)
```

Comment:
This seems to be a constructor for `FeatureVector` that's implemented as a method of another class. Any reason not to move it across?